service/worker/batcher/workflow.go

```diff
@@ -20,7 +20,328 @@
+		//  we can't use list API because terminate / reset will mutate the result.
+		resp, err := batcher.svcClient.ScanWorkflowExecutions(ctx, &shared.ListWorkflowExecutionsRequest{
+			PageSize:      common.Int32Ptr(int32(pageSize)),
+			Domain:        common.StringPtr(batchParams.DomainName),
+			NextPageToken: hbd.PageToken,
+			Query:         common.StringPtr(batchParams.QueryCondition),
+		})
+		if err != nil {
+			return err
+		}
+
+		// clear all counters
+		atomic.StoreInt32(&succCount, 0)
+		atomic.StoreInt32(&errCount, 0)
+
+		// send all tasks
+		for _, wf := range resp.Executions {
+			taskCh <- taskDetail{
+				execution: *wf.Execution,
+				attempts:  0,
+				hbd:       hbd,
+			}
+		}
+		batchCount := len(resp.Executions)
+
+		// wait for counters indicate this batch is done
+	Loop:
+		for {
+			select {
+			case <-time.Tick(time.Second):
+				if int32(batchCount) == atomic.LoadInt32(&succCount)+atomic.LoadInt32(&errCount) {
+					break Loop
+				}
+			case <-ctx.Done():
+				return nil
+			}
+		}
+
+		hbd.CurrentPage++
+		hbd.PageToken = resp.NextPageToken
+		hbd.SuccessCount += atomic.LoadInt32(&succCount)
+		hbd.ErrorCount += atomic.LoadInt32(&errCount)
+		activity.RecordHeartbeat(ctx, hbd)
+
+		if len(hbd.PageToken) == 0 {
+			break
+		}
+	}
+
+	return nil
+}
+
+func startTaskProcessor(
+	ctx context.Context, batchParams BatchParams, taskCh chan taskDetail,
+	limiter *rate.Limiter, doneCount, errCount *int32) {
+
+	for {
+		select {
+		case task := <-taskCh:
+			if isDone(ctx) {
+				return
+			}
+			var err error
+
+			switch batchParams.BatchType {
+			case BatchTypeTerminate:
+				err = processOneTerminateTask(ctx, limiter, task, batchParams)
+			}
+			if err != nil {
+				getActivityLogger(ctx).Error("Failed to process task", tag.Error(err))
+				// put back to the channel if less than attemptsOnError
+				task.attempts++
+				if task.attempts >= batchParams.AttemptsOnRetryableError {
+					atomic.AddInt32(errCount, 1)
+				} else {
+					taskCh <- task
+				}
+			} else {
+				atomic.AddInt32(doneCount, 1)
+			}
+		}
+	}
+}
+
+func processOneTerminateTask(ctx context.Context, limiter *rate.Limiter, task taskDetail, param BatchParams) error {
+	batcher := ctx.Value(batcherContextKey).(*Batcher)
+
+	wfs := []shared.WorkflowExecution{task.execution}
+	for len(wfs) > 0 {
+		wf := wfs[0]
+
+		err := limiter.Wait(ctx)
+		if err != nil {
+			getActivityLogger(ctx).Error("Failed to wait for rateLimiter", tag.Error(err))
+		}
+
+		newCtx, cancel := context.WithDeadline(ctx, time.Now().Add(rpcTimeout))
+		err = batcher.svcClient.TerminateWorkflowExecution(newCtx, &shared.TerminateWorkflowExecutionRequest{
+			Domain:            common.StringPtr(param.DomainName),
+			WorkflowExecution: &wf,
+			Reason:            common.StringPtr(param.Reason),
+			Identity:          common.StringPtr(batchWFTypeName),
+		})
+		cancel()
+		if err != nil {
+			_, ok := err.(*shared.EntityNotExistsError)
+			if !ok {
```

Comment:
Shouldn't we return the error in either case? If parent terminate fails even for a retryable error we shouldnt try to terminate children.. We should try to terminate parent first..

Reply:
That's because we need to continue on terminating all childWFs in this case. Do you have any idea to implement if we returning early here?